File: ios/Classes/AwsIvsFlutterPlugin.swift
import Flutter
import UIKit
import AmazonIVSPlayer

public class AwsIvsFlutterPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var currentPlayer: IVSPlayer?
    private var playerView: IVSPlayerView?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "aws_ivs_flutter", binaryMessenger: registrar.messenger())
        let instance = AwsIvsFlutterPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Register platform view factory
        let factory = AwsIvsPlayerViewFactory(messenger: registrar.messenger())
        registrar.register(factory, withId: "aws_ivs_player_view")
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "createPlayer":
            createPlayer(call: call, result: result)
            
        case "play":
            playVideo(result: result)
            
        case "pause":
            pauseVideo(result: result)
            
        case "stop":
            stopVideo(result: result)
            
        case "dispose":
            disposePlayer(result: result)
            
        case "loadStream":
            loadStream(call: call, result: result)
            
        case "setVolume":
            setVolume(call: call, result: result)
            
        case "getPlayerState":
            getPlayerState(result: result)
            
        case "getDuration":
            getDuration(result: result)
            
        case "getPosition":
            getPosition(result: result)
            
        case "seekTo":
            seekTo(call: call, result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func createPlayer(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let streamUrl = args["streamUrl"] as? String,
              !streamUrl.isEmpty else {
            result(FlutterError(code: "INVALID_URL", message: "Stream URL cannot be null or empty", details: nil))
            return
        }
        
        do {
            // Create IVS Player
            currentPlayer = IVSPlayer()
            
            // Load the stream
            if let url = URL(string: streamUrl) {
                currentPlayer?.load(url)
                print("AwsIvsFlutterPlugin: Player created successfully for URL: \(streamUrl)")
                result("Player created successfully")
            } else {
                result(FlutterError(code: "INVALID_URL", message: "Invalid stream URL format", details: nil))
            }
        } catch {
            print("AwsIvsFlutterPlugin: Error creating player: \(error)")
            result(FlutterError(code: "PLAYER_CREATION_ERROR", message: "Failed to create player: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func playVideo(result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        do {
            player.play()
            print("AwsIvsFlutterPlugin: Playback started")
            result("Player started")
        } catch {
            print("AwsIvsFlutterPlugin: Error starting playback: \(error)")
            result(FlutterError(code: "PLAYBACK_ERROR", message: "Failed to start playback: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func pauseVideo(result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        do {
            player.pause()
            print("AwsIvsFlutterPlugin: Playback paused")
            result("Player paused")
        } catch {
            print("AwsIvsFlutterPlugin: Error pausing playback: \(error)")
            result(FlutterError(code: "PAUSE_ERROR", message: "Failed to pause playback: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func stopVideo(result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        do {
            player.pause()
            print("AwsIvsFlutterPlugin: Playback stopped")
            result("Player stopped")
        } catch {
            print("AwsIvsFlutterPlugin: Error stopping playback: \(error)")
            result(FlutterError(code: "STOP_ERROR", message: "Failed to stop playback: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func disposePlayer(result: @escaping FlutterResult) {
        do {
            currentPlayer?.pause()
            currentPlayer = nil
            playerView = nil
            print("AwsIvsFlutterPlugin: Player disposed")
            result("Player disposed")
        } catch {
            print("AwsIvsFlutterPlugin: Error disposing player: \(error)")
            result(FlutterError(code: "DISPOSE_ERROR", message: "Failed to dispose player: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func loadStream(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let streamUrl = args["streamUrl"] as? String,
              !streamUrl.isEmpty else {
            result(FlutterError(code: "INVALID_URL", message: "Stream URL cannot be null or empty", details: nil))
            return
        }
        
        do {
            if currentPlayer == nil {
                currentPlayer = IVSPlayer()
            }
            
            if let url = URL(string: streamUrl) {
                currentPlayer?.load(url)
                print("AwsIvsFlutterPlugin: Stream loaded: \(streamUrl)")
                result("Stream loaded successfully")
            } else {
                result(FlutterError(code: "INVALID_URL", message: "Invalid stream URL format", details: nil))
            }
        } catch {
            print("AwsIvsFlutterPlugin: Error loading stream: \(error)")
            result(FlutterError(code: "LOAD_ERROR", message: "Failed to load stream: \(error.localizedDescription)", details: nil))
        }
    }
    
    private func setVolume(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        guard let args = call.arguments as? [String: Any],
              let volume = args["volume"] as? Double else {
            result(FlutterError(code: "INVALID_VOLUME", message: "Invalid volume value", details: nil))
            return
        }
        
        let clampedVolume = max(0.0, min(1.0, volume))
        player.volume = Float(clampedVolume)
        print("AwsIvsFlutterPlugin: Volume set to: \(clampedVolume)")
        result("Volume set successfully")
    }
    
    private func getPlayerState(result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        let stateString: String
        switch player.state {
        case .idle:
            stateString = "IDLE"
        case .ready:
            stateString = "READY"
        case .buffering:
            stateString = "BUFFERING"
        case .playing:
            stateString = "PLAYING"
        case .ended:
            stateString = "ENDED"
        @unknown default:
            stateString = "UNKNOWN"
        }
        
        result(stateString)
    }
    
    private func getDuration(result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        let duration = player.duration.seconds * 1000 // Convert to milliseconds
        result(Int64(duration))
    }
    
    private func getPosition(result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        let position = player.position.seconds * 1000 // Convert to milliseconds
        result(Int64(position))
    }
    
    private func seekTo(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let player = currentPlayer else {
            result(FlutterError(code: "PLAYER_NOT_INITIALIZED", message: "Player not initialized", details: nil))
            return
        }
        
        guard let args = call.arguments as? [String: Any],
              let positionMs = args["position"] as? Int64 else {
            result(FlutterError(code: "INVALID_POSITION", message: "Invalid seek position", details: nil))
            return
        }
        
        let positionSeconds = CMTime(seconds: Double(positionMs) / 1000.0, preferredTimescale: 1000)
        player.seek(to: positionSeconds)
        print("AwsIvsFlutterPlugin: Seeking to position: \(positionMs) ms")
        result("Seek completed")
    }
}

// MARK: - Platform View Factory
class AwsIvsPlayerViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        return AwsIvsPlayerViewController(frame: frame, viewIdentifier: viewId, arguments: args, binaryMessenger: messenger)
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

// MARK: - Platform View Controller
class AwsIvsPlayerViewController: NSObject, FlutterPlatformView {
    private var _view: UIView
    private var player: IVSPlayer?
    private var playerView: IVSPlayerView?
    
    init(frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?, binaryMessenger messenger: FlutterBinaryMessenger) {
        _view = UIView()
        super.init()
        
        createPlayer(args: args)
    }
    
    func view() -> UIView {
        return _view
    }
    
    private func createPlayer(args: Any?) {
        do {
            // Create IVS Player
            player = IVSPlayer()
            player?.delegate = self
            
            // Create IVS Player View
            playerView = IVSPlayerView()
            playerView?.player = player
            
            // Configure player view
            if let playerView = playerView {
                playerView.videoGravity = .resizeAspect
                playerView.translatesAutoresizingMaskIntoConstraints = false
                _view.addSubview(playerView)
                
                // Add constraints
                NSLayoutConstraint.activate([
                    playerView.topAnchor.constraint(equalTo: _view.topAnchor),
                    playerView.leadingAnchor.constraint(equalTo: _view.leadingAnchor),
                    playerView.trailingAnchor.constraint(equalTo: _view.trailingAnchor),
                    playerView.bottomAnchor.constraint(equalTo: _view.bottomAnchor)
                ])
            }
            
            // Load stream URL if provided in creation parameters
            if let creationParams = args as? [String: Any],
               let streamUrl = creationParams["streamUrl"] as? String,
               !streamUrl.isEmpty,
               let url = URL(string: streamUrl) {
                player?.load(url)
                print("AwsIvsPlayerViewController: Stream loaded from creation params: \(streamUrl)")
            }
            
            print("AwsIvsPlayerViewController: Player view initialized successfully")
            
        } catch {
            print("AwsIvsPlayerViewController: Error setting up player: \(error)")
        }
    }
}

// MARK: - IVS Player Delegate
extension AwsIvsPlayerViewController: IVSPlayerDelegate {
    func player(_ player: IVSPlayer, didChangeState state: IVSPlayerState) {
        let stateString: String
        switch state {
        case .idle:
            stateString = "IDLE"
        case .ready:
            stateString = "READY"
        case .buffering:
            stateString = "BUFFERING"
        case .playing:
            stateString = "PLAYING"
        case .ended:
            stateString = "ENDED"
        @unknown default:
            stateString = "UNKNOWN"
        }
        print("AwsIvsPlayerViewController: Player state changed: \(stateString)")
    }
    
    func player(_ player: IVSPlayer, didFailWithError error: Error) {
        print("AwsIvsPlayerViewController: Player error: \(error.localizedDescription)")
    }
    
    func player(_ player: IVSPlayer, didChangeVideoSize videoSize: CGSize) {
        print("AwsIvsPlayerViewController: Video size changed: \(videoSize)")
    }
    
    func player(_ player: IVSPlayer, didChangeDuration duration: CMTime) {
        print("AwsIvsPlayerViewController: Duration changed: \(duration.seconds) seconds")
    }
    
    func player(_ player: IVSPlayer, didOutputMetadataTimedMetadataGroup metadataGroup: AVTimedMetadataGroup) {
        print("AwsIvsPlayerViewController: Metadata received")
    }
    
    func playerWillRebuffer(_ player: IVSPlayer) {
        print("AwsIvsPlayerViewController: Player will rebuffer")
    }
    
    func player(_ player: IVSPlayer, didSeekTo time: CMTime) {
        print("AwsIvsPlayerViewController: Seek completed to: \(time.seconds) seconds")
    }
}

File: android/src/main/kotlin/com/example/aws_ivs_flutter/AwsIvsFlutterPlugin.kt
package com.example.aws_ivs_flutter

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.NonNull
import com.amazonaws.ivs.player.*
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class AwsIvsFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var context: Context? = null
    private var currentPlayer: Player? = null

    companion object {
        private const val TAG = "AwsIvsFlutterPlugin"
        private const val CHANNEL_NAME = "aws_ivs_flutter"
        private const val VIEW_TYPE = "aws_ivs_player_view"
        @JvmStatic
        var videoWidth: Int = 0
        @JvmStatic
        var videoHeight: Int = 0
    }

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
        flutterPluginBinding
            .platformViewRegistry
            .registerViewFactory(VIEW_TYPE, AwsIvsPlayerViewFactory(context!!))
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "createPlayer" -> createPlayer(call, result)
            "play" -> play(result)
            "pause" -> pause(result)
            "stop" -> stop(result)
            "loadStream" -> loadStream(call, result)
            "setVolume" -> setVolume(call, result)
            "getPlayerState" -> getPlayerState(result)
            "getDuration" -> getDuration(result)
            "getPosition" -> getPosition(result)
            "seekTo" -> seekTo(call, result)
            "dispose" -> dispose(result)
            "getVideoWidth" -> {
                result.success(AwsIvsPlayerView.videoWidth.takeIf { it > 0 } ?: 0)
            }
            "getVideoHeight" -> {
                result.success(AwsIvsPlayerView.videoHeight.takeIf { it > 0 } ?: 0)
            }
            else -> result.notImplemented()
        }
    }

    private fun createPlayer(call: MethodCall, result: Result) {
        try {
            val streamUrl = call.argument<String>("streamUrl")
            if (streamUrl.isNullOrEmpty()) {
                result.error("INVALID_URL", "Stream URL cannot be null or empty", null)
                return
            }
            currentPlayer = Player.Factory.create(context!!)
            val uri = Uri.parse(streamUrl)
            currentPlayer?.load(uri)
            Log.d(TAG, "Player created successfully for URL: $streamUrl")
            result.success("Player created successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Error creating player", e)
            result.error("PLAYER_CREATION_ERROR", "Failed to create player: ${e.message}", null)
        }
    }

    private fun loadStream(call: MethodCall, result: Result) {
        try {
            val streamUrl = call.argument<String>("streamUrl")
            if (streamUrl.isNullOrEmpty()) {
                result.error("INVALID_URL", "Stream URL cannot be null or empty", null)
                return
            }

            if (currentPlayer == null) {
                currentPlayer = Player.Factory.create(context!!)
            }

            val uri = Uri.parse(streamUrl)
            currentPlayer?.load(uri)

            Log.d(TAG, "Stream loaded: $streamUrl")
            result.success("Stream loaded successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Error loading stream", e)
            result.error("LOAD_ERROR", "Failed to load stream: ${e.message}", null)
        }
    }

    private fun play(result: Result) {
        try {
            currentPlayer?.play()
            result.success("Playing")
        } catch (e: Exception) {
            Log.e(TAG, "Error starting playback", e)
            result.error("PLAY_ERROR", "Error playing: ${e.message}", null)
        }
    }

    private fun pause(result: Result) {
        try {
            currentPlayer?.pause()
            result.success("Paused")
        } catch (e: Exception) {
            Log.e(TAG, "Error pausing playback", e)
            result.error("PAUSE_ERROR", "Error pausing: ${e.message}", null)
        }
    }

    private fun stop(result: Result) {
        try {
            currentPlayer?.pause()
            result.success("Stopped")
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping playback", e)
            result.error("STOP_ERROR", "Error stopping: ${e.message}", null)
        }
    }

    private fun setVolume(call: MethodCall, result: Result) {
        try {
            val volume = call.argument<Double>("volume") ?: 1.0
            currentPlayer?.volume = volume.toFloat()
            result.success("Volume set to $volume")
        } catch (e: Exception) {
            Log.e(TAG, "Error setting volume", e)
            result.error("SET_VOLUME_ERROR", "Error setting volume: ${e.message}", null)
        }
    }

    private fun getPlayerState(result: Result) {
        try {
            val state = when (currentPlayer?.state) {
                Player.State.IDLE -> "IDLE"
                Player.State.READY -> "READY"
                Player.State.BUFFERING -> "BUFFERING"
                Player.State.PLAYING -> "PLAYING"
                Player.State.ENDED -> "ENDED"
                else -> "UNKNOWN"
            }
            result.success(state)
        } catch (e: Exception) {
            Log.e(TAG, "Error getting player state", e)
            result.error("GET_STATE_ERROR", "Error getting player state: ${e.message}", null)
        }
    }

    private fun getDuration(result: Result) {
        try {
            val duration = currentPlayer?.duration ?: 0L
            result.success(duration)
        } catch (e: Exception) {
            Log.e(TAG, "Error getting duration", e)
            result.error("GET_DURATION_ERROR", "Error getting duration: ${e.message}", null)
        }
    }

    private fun getPosition(result: Result) {
        try {
            val position = currentPlayer?.position ?: 0L
            result.success(position)
        } catch (e: Exception) {
            Log.e(TAG, "Error getting position", e)
            result.error("GET_POSITION_ERROR", "Error getting position: ${e.message}", null)
        }
    }

    private fun seekTo(call: MethodCall, result: Result) {
        try {
            val position = call.argument<Long>("position") ?: 0L
            currentPlayer?.seekTo(position)
            result.success("Seeked to $position")
        } catch (e: Exception) {
            Log.e(TAG, "Error seeking", e)
            result.error("SEEK_ERROR", "Error seeking: ${e.message}", null)
        }
    }

    private fun dispose(result: Result) {
        try {
            currentPlayer?.release()
            currentPlayer = null
            result.success("Player disposed")
        } catch (e: Exception) {
            Log.e(TAG, "Error disposing player", e)
            result.error("DISPOSE_ERROR", "Error disposing player: ${e.message}", null)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        try {
            currentPlayer?.release()
            currentPlayer = null
        } catch (e: Exception) {
            Log.e(TAG, "Error during cleanup", e)
        }
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

class AwsIvsPlayerViewFactory(private val context: Context) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return AwsIvsPlayerView(context, viewId, args)
    }
}

class AwsIvsPlayerView(
    private val context: Context,
    private val viewId: Int,
    private val creationParams: Any?
) : PlatformView, Player.Listener() {

    private val frameLayout: FrameLayout = FrameLayout(context)
    private var player: Player? = null
    private var playerView: PlayerView? = null

    companion object {
        private const val TAG = "AwsIvsPlayerView"
        @JvmStatic
        var videoWidth: Int = 0
        @JvmStatic
        var videoHeight: Int = 0
    }

    init {
        setupPlayer()
    }

    private fun setupPlayer() {
        try {
            // Initialize IVS Player
            player = Player.Factory.create(context)
            player?.addListener(this)
            player?.volume = 1.0f
            // Create PlayerView with proper constructor
            playerView = PlayerView(context).apply {
                resizeMode = ResizeMode.FIT
                controlsEnabled = false
                // Set layout params after creation
                this.layoutParams = FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.MATCH_PARENT,
                    FrameLayout.LayoutParams.MATCH_PARENT
                )
                // Set player using reflection or direct property access
                try {
                    val playerField = this.javaClass.getDeclaredField("player")
                    playerField.isAccessible = true
                    playerField.set(this, this@AwsIvsPlayerView.player)
                } catch (e: Exception) {
                    Log.e(TAG, "Could not set player via reflection", e)
                }
            }

            frameLayout.addView(playerView)

            // Load stream URL if provided in creation parameters
            val params = creationParams as? Map<String, Any>
            val streamUrl = params?.get("streamUrl") as? String
            if (!streamUrl.isNullOrEmpty()) {
                loadStream(streamUrl)
            }

            Log.d(TAG, "Player view initialized successfully")

        } catch (e: Exception) {
            Log.e(TAG, "Error setting up player", e)
        }
    }

    private fun loadStream(streamUrl: String) {
        try {
            if (player == null) {
                player = Player.Factory.create(context)
                player?.addListener(this)
                player?.volume = 1.0f
                playerView?.let { frameLayout.removeView(it) }

                playerView = PlayerView(context).apply {
                    resizeMode = ResizeMode.FIT
                    controlsEnabled = false
                    layoutParams = FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.MATCH_PARENT,
                        FrameLayout.LayoutParams.MATCH_PARENT
                    )
                    try {
                        val playerField = javaClass.getDeclaredField("player")
                        playerField.isAccessible = true
                        playerField.set(this, this@AwsIvsPlayerView.player)
                    } catch (e: Exception) {
                        Log.e(TAG, "Could not set player via reflection", e)
                    }
                }

                frameLayout.addView(playerView)
            }

            val uri = Uri.parse(streamUrl)
            player?.load(uri)
            player?.play()
            Log.d(TAG, "Stream loaded: $streamUrl")
        } catch (e: Exception) {
            Log.e(TAG, "Error loading stream: $streamUrl", e)
        }
    }

    override fun getView(): View {
        return frameLayout
    }

    override fun dispose() {
        try {
            player?.removeListener(this)
            player?.release()
            player = null
            playerView = null
            Log.d(TAG, "Player disposed")
        } catch (e: Exception) {
            Log.e(TAG, "Error disposing player", e)
        }
    }

    // Player.Listener implementation
    override fun onCue(cue: Cue) {
        Log.d(TAG, "Cue received")
    }

    override fun onDurationChanged(duration: Long) {
        Log.d(TAG, "Duration changed: $duration ms")
    }

    override fun onError(exception: PlayerException) {
        Log.e(TAG, "Player error: ${exception.errorMessage}", exception)
    }

    override fun onQualityChanged(quality: Quality) {
        Log.d(TAG, "Quality changed: ${quality.name}")
    }

    override fun onRebuffering() {
        Log.d(TAG, "Player rebuffering")
    }

    override fun onSeekCompleted(position: Long) {
        Log.d(TAG, "Seek completed: $position ms")
    }

    override fun onVideoSizeChanged(width: Int, height: Int) {
        Log.d(TAG, "Video size changed: ${width}x${height}")
        videoWidth = width
        videoHeight = height
    }

    override fun onStateChanged(state: Player.State) {
        val stateString = when (state) {
            Player.State.IDLE -> "IDLE"
            Player.State.READY -> "READY"
            Player.State.BUFFERING -> "BUFFERING"
            Player.State.PLAYING -> "PLAYING"
            Player.State.ENDED -> "ENDED"
        }
        Log.d(TAG, "Player state changed: $stateString")
    }
}